// grakn-client/concept/Thing/Impl/RelationImpl.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("grakn-client/concept/Thing/Impl/RelationImpl", ["require", "exports", "grakn-client/concept/Thing/Impl/ThingImpl", "grakn-client/concept/Concept"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RemoteRelationImpl = exports.RelationImpl = void 0;
    const ThingImpl_1 = require("grakn-client/concept/Thing/Impl/ThingImpl");
    const Concept_1 = require("grakn-client/concept/Concept");
    class RelationImpl extends ThingImpl_1.ThingImpl {
        constructor(iid) {
            super(iid);
        }
        static of(protoThing) {
            return new RelationImpl(protoThing.getIid_asB64());
        }
        asRemote(transaction) {
            return new RemoteRelationImpl(transaction, this.getIID());
        }
    }
    exports.RelationImpl = RelationImpl;
    class RemoteRelationImpl extends ThingImpl_1.RemoteThingImpl {
        constructor(transaction, iid) {
            super(transaction, iid);
        }
        asRemote(transaction) {
            return new RemoteRelationImpl(transaction, this.getIID());
        }
        getType() {
            throw "As yet unimplemented";
        }
        getPlayersByRoleType() {
            throw "Not implemented";
        }
        getPlayers(roleTypes) {
            return new Concept_1.QueryIterator();
        }
        addPlayer(roleType, player) {
            throw "Not implemented";
        }
        removePlayer(roleType, player) {
            throw "Not implemented";
        }
    }
    exports.RemoteRelationImpl = RemoteRelationImpl;
});
