// grakn-client/concept/Thing/Impl/ThingImpl.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("grakn-client/concept/Thing/Impl/ThingImpl", ["require", "exports", "grakn-client/concept/Concept"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RemoteThingImpl = exports.ThingImpl = void 0;
    const Concept_1 = require("grakn-client/concept/Concept");
    class ThingImpl {
        // TODO: all error messages should be extracted into ErrorMessage class or namespace
        constructor(iid) {
            if (!iid) {
                throw "IID Missing";
            }
            this._iid = iid;
        }
        getIID() {
            return this._iid;
        }
        isRemote() {
            return false;
        }
        toString() {
            return `${ThingImpl.name}[iid:${this._iid}]`;
        }
    }
    exports.ThingImpl = ThingImpl;
    class RemoteThingImpl {
        constructor(transaction, iid) {
            if (!transaction)
                throw "Transaction Missing";
            if (!iid)
                throw "IID Missing";
            this._iid = iid;
            this._transaction = transaction;
        }
        getType() {
            throw "Not implemented yet";
        }
        isInferred() {
            throw "Not implemented yet";
        }
        delete() {
            throw "Not implemented yet";
        }
        getIID() {
            return this._iid;
        }
        isDeleted() {
            return false;
        }
        isRemote() {
            return true;
        }
        getHas(onlyKey) {
            return new Concept_1.QueryIterator();
        }
        getPlays() {
            return new Concept_1.QueryIterator();
        }
        getRelations(roleTypes) {
            return new Concept_1.QueryIterator();
        }
        setHas(attribute) {
            throw "Not implemented yet";
        }
        unsetHas(attribute) {
            throw "Not implemented yet";
        }
        get transaction() {
            return this._transaction;
        }
        toString() {
            return `${RemoteThingImpl.name}[iid:${this._iid}]`;
        }
    }
    exports.RemoteThingImpl = RemoteThingImpl;
});
