// grakn-client/concept/Type/Impl/RelationTypeImpl.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("grakn-client/concept/Type/Impl/RelationTypeImpl", ["require", "exports", "grakn-client/concept/Type/Impl/ThingTypeImpl"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RemoteRelationTypeImpl = exports.RelationTypeImpl = void 0;
    const ThingTypeImpl_1 = require("grakn-client/concept/Type/Impl/ThingTypeImpl");
    class RelationTypeImpl extends ThingTypeImpl_1.ThingTypeImpl {
        constructor(label, isRoot) {
            super(label, isRoot);
        }
        static of(typeProto) {
            return new RelationTypeImpl(typeProto.getLabel(), typeProto.getRoot());
        }
        asRemote(transaction) {
            return new RemoteRelationTypeImpl(transaction, this.getLabel(), this.isRoot());
        }
    }
    exports.RelationTypeImpl = RelationTypeImpl;
    class RemoteRelationTypeImpl extends ThingTypeImpl_1.RemoteThingTypeImpl {
        constructor(transaction, label, isRoot) {
            super(transaction, label, isRoot);
        }
        getInstances() {
            throw "Not yet implemented";
        }
        asRemote(transaction) {
            return new RemoteRelationTypeImpl(transaction, this.getLabel(), this.isRoot());
        }
        getSupertype() {
            throw "Not yet implemented";
        }
        getSupertypes() {
            throw "Not yet implemented";
        }
        getSubtypes() {
            throw "Not yet implemented";
        }
        setSupertype(superRelationType) {
            throw "Not yet implemented";
        }
        create() {
            throw "As yet unimplemented";
        }
        getRelates(roleLabel) {
            throw "Not yet implemented";
        }
        setRelates(roleLabel, overriddenLabel) {
            throw "Not yet implemented";
        }
        unsetRelates(roleLabel) {
            throw "Not yet implemented";
        }
    }
    exports.RemoteRelationTypeImpl = RemoteRelationTypeImpl;
});
