// grakn-client/concept/Type/Impl/RoleTypeImpl.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("grakn-client/concept/Type/Impl/RoleTypeImpl", ["require", "exports", "grakn-client/concept/Type/Impl/ThingTypeImpl"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RemoteRoleTypeImpl = exports.RoleTypeImpl = void 0;
    const ThingTypeImpl_1 = require("grakn-client/concept/Type/Impl/ThingTypeImpl");
    class RoleTypeImpl extends ThingTypeImpl_1.ThingTypeImpl {
        constructor(label, scope, isRoot) {
            super(label, isRoot);
            this._scope = scope;
        }
        static of(typeProto) {
            return new RoleTypeImpl(typeProto.getLabel(), typeProto.getScope(), typeProto.getRoot());
        }
        getScope() {
            return this._scope;
        }
        asRemote(transaction) {
            return new RemoteRoleTypeImpl(transaction, this.getLabel(), this.getScope(), this.isRoot());
        }
        toString() {
            return `${RoleTypeImpl.name}[label: ${this._scope ? `${this._scope}:${this.getLabel()}` : this.getLabel()}]`;
        }
    }
    exports.RoleTypeImpl = RoleTypeImpl;
    class RemoteRoleTypeImpl extends ThingTypeImpl_1.RemoteThingTypeImpl {
        constructor(transaction, label, scope, isRoot) {
            super(transaction, label, isRoot);
            this._scope = scope;
        }
        getSupertype() {
            throw "Not yet implemented";
        }
        getSupertypes() {
            throw "Not yet implemented";
        }
        getSubtypes() {
            throw "Not yet implemented";
        }
        getScope() {
            return this._scope;
        }
        asRemote(transaction) {
            return new RemoteRoleTypeImpl(transaction, this.getLabel(), this._scope, this.isRoot());
        }
        getRelation() {
            throw "Not yet implemented";
        }
        getRelations() {
            throw "Not yet implemented";
        }
        getPlayers() {
            throw "Not yet implemented";
        }
        toString() {
            return `${RemoteRoleTypeImpl.name}[label: ${this._scope ? `${this._scope}:${this.getLabel()}` : this.getLabel()}]`;
        }
    }
    exports.RemoteRoleTypeImpl = RemoteRoleTypeImpl;
});
