// grakn-client/concept/Type/Impl/RuleImpl.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("grakn-client/concept/Type/Impl/RuleImpl", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RemoteRuleImpl = exports.RuleImpl = void 0;
    class RuleImpl {
        constructor(label, when, then) {
            this._label = label;
            this._when = when;
            this._then = then;
        }
        static of(ruleProto) {
            return new RuleImpl(ruleProto.getLabel(), ruleProto.getWhen(), ruleProto.getThen());
        }
        getLabel() {
            return this._label;
        }
        getThen() {
            return this._then;
        }
        getWhen() {
            return this._when;
        }
        asRemote(transaction) {
            return new RemoteRuleImpl(transaction, this.getLabel(), this.getWhen(), this.getThen());
        }
        toString() {
            return `${RuleImpl.name}[label:${this._label}]`;
        }
    }
    exports.RuleImpl = RuleImpl;
    class RemoteRuleImpl {
        constructor(transaction, label, when, then) {
            this._transaction = transaction;
            this._label = label;
            this._when = when;
            this._then = then;
        }
        getLabel() {
            return this._label;
        }
        getThen() {
            return this._then;
        }
        getWhen() {
            return this._when;
        }
        setLabel(label) {
            this._label = label;
            throw "Not yet implemented"; // TODO: rpc call
        }
        delete() {
            throw "Not yet implemented";
        }
        isDeleted() {
            throw "Not yet applicable";
        }
        asRemote(transaction) {
            return new RemoteRuleImpl(transaction, this.getLabel(), this.getWhen(), this.getThen());
        }
        get transaction() {
            return this._transaction;
        }
    }
    exports.RemoteRuleImpl = RemoteRuleImpl;
});
