// grakn-client/concept/Type/Impl/ThingTypeImpl.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("grakn-client/concept/Type/Impl/ThingTypeImpl", ["require", "exports", "grakn-client/concept/Type/Impl/TypeImpl", "graknlabs-grpc-protocol/protobuf/concept_pb", "assert"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RemoteThingTypeImpl = exports.ThingTypeImpl = void 0;
    const TypeImpl_1 = require("grakn-client/concept/Type/Impl/TypeImpl");
    const concept_pb_1 = __importDefault(require("graknlabs-grpc-protocol/protobuf/concept_pb"));
    var TypeProto = concept_pb_1.default.Type;
    const assert_1 = __importDefault(require("assert"));
    class ThingTypeImpl extends TypeImpl_1.TypeImpl {
        constructor(label, isRoot) {
            super(label, isRoot);
        }
        asRemote(transaction) {
            return new RemoteThingTypeImpl(transaction, this.getLabel(), this.isRoot());
        }
        static of(typeProto) {
            switch (typeProto.getEncoding()) {
                case TypeProto.ENCODING.ENTITY_TYPE:
                    throw "no"; // TODO: resolve circular ref
                // return EntityTypeImpl.of(typeProto);
                case TypeProto.ENCODING.RELATION_TYPE:
                    throw "no"; // TODO: resolve circular ref
                // return RelationTypeImpl.of(typeProto);
                case TypeProto.ENCODING.ATTRIBUTE_TYPE:
                    throw "no"; // TODO: resolve circular ref
                // return AttributeTypeImpl.of(typeProto);
                case TypeProto.ENCODING.THING_TYPE:
                    assert_1.default(typeProto.getRoot());
                    return new ThingTypeImpl(typeProto.getLabel(), typeProto.getRoot());
                default:
                    throw "Bad encoding";
            }
        }
    }
    exports.ThingTypeImpl = ThingTypeImpl;
    class RemoteThingTypeImpl extends TypeImpl_1.RemoteTypeImpl {
        constructor(transaction, label, isRoot) {
            super(transaction, label, isRoot);
        }
        getSupertype() {
            throw "Behaviour not yet implemented";
        }
        getSupertypes() {
            throw "Not implemented yet";
        }
        getSubtypes() {
            throw "Not implemented yet";
        }
        getInstances() {
            const request = new concept_pb_1.default.Type.Req()
                .setThingTypeGetInstancesReq(new concept_pb_1.default.ThingType.GetInstances.Req());
            return this.thingStream(request, res => res.getThingTypeGetInstancesRes().getThingList());
        }
        setAbstract() {
            throw "Not implemented yet";
        }
        unsetAbstract() {
            throw "Not implemented yet";
        }
        getPlays() {
            throw "Not implemented yet";
        }
        getOwns(keysOnly) {
            throw "Not implemented yet";
        }
        setOwns(attributeType, isKey, otherType) {
        }
        setPlays(role, overriddenType) {
            throw "Not implemented yet";
        }
        unsetOwns(attributeType) {
            throw "Not implemented yet";
        }
        unsetPlays(role) {
            throw "Not implemented yet";
        }
        asRemote(transaction) {
            return new RemoteThingTypeImpl(transaction, this.getLabel(), this.isRoot());
        }
    }
    exports.RemoteThingTypeImpl = RemoteThingTypeImpl;
});
