// grakn-client/GraknOptions.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("grakn-client/GraknOptions", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.GraknOptions = void 0;
    class GraknOptions {
        constructor() {
            this._infer = null;
            this._explain = null;
            this._batchSize = null;
        }
        infer() {
            return this._infer;
        }
        set_infer(infer) {
            this._infer = infer;
            return this;
        }
        explain() {
            return this._explain;
        }
        set_explain(explain) {
            this._explain = explain;
            return this;
        }
        batchSize() {
            return this._batchSize;
        }
        set_batchSize(batchSize) {
            if (batchSize < 1) {
                throw "Attempted to set nonpositive batch size.";
            }
            this._batchSize = batchSize;
            return this;
        }
    }
    exports.GraknOptions = GraknOptions;
});
